middleware_test.go

```diff
@@ -136,6 +136,23 @@ func TestMiddlewareSubrouter(t *testing.T) {
 	}
 }
 
+func TestMiddlewareDuplicateSubrouter(t *testing.T) {
+	router := NewRouter()
+	router.PathPrefix("/sub").Subrouter() // Deliberately create a duplicate router
```

Comment:
@elithrar I'm not sure we agree that this is a user error, as you suggested in your next comment.  I believe a subrouter is something with a path prefix and possibly some middleware.  It's not obvious to me that subrouters should be mutually exclusive unless that is implied by the possibility of having a NotFound handler (specifically by the resultant ambiguity about which NotFound or MethodNotAllowedHandler to use in the case of overlapping subroutes).  Also, while we can recognize identical subroute segments, I think recognizing the full set of possibly overlapping routes would add considerable complexity.  The feature of allowing reuse of middleware between arbitrary groups of routes does seem a valuable one.  I've updated the test case name and behavior to match the simpler example provided by @tkinz27.  The problem is not specific to duplication -- I was just trying to come up with the simplest possible example -- but I think the overlapping subrouter and path case is clearer.

Reply:
Mux doesn't really support multiple handlers, whether explicit or as subrouters, at the same path. I actually think instead of overwriting the old one it should probably panic, since it's basically a configuration error that can lead to confusion and unexpected results.

Reply:
I updated my test case so it wasn't simply a case of duplication.  That was just a subset of the cases where the middleware isn't run.  The current test case setup looks like this:

```
	router.PathPrefix("/x").Subrouter()
	router.HandleFunc("/x", dummyHandler).Methods("GET")
```
@kisielk  In this case, do you feel the second line should panic?

Reply:
That's up for debate, but IMO you should put all paths rooted at a subrouter on the subrouter itself to avoid accidental duplication.